src/WebJobs.Script.WebHost/Diagnostics/FunctionInstanceLogger.cs

```diff
@@ -131,12 +132,11 @@ public async Task AddAsync(FunctionInstanceLogEntry item, CancellationToken canc
 
         private FunctionMetadata GetFunctionMetadata(string functionName)
         {
-            FunctionMetadata GetMetadataFromCollection(IEnumerable<FunctionMetadata> functions)
+            return _functionMetadataMap.GetOrAdd(functionName, s =>
             {
-                return functions.FirstOrDefault(p => Utility.FunctionNamesMatch(p.Name, functionName));
-            }
-
-            return GetMetadataFromCollection(_metadataManager.GetFunctionMetadata());
+                var functions = _metadataManager.GetFunctionMetadata();
```

Comment:
Good point - the metadata manager is a WebHost level service, and internally it can reset itself and we'd miss that here. What I can do is promote this the manager interface by adding a TryGetFunctionMetadata method. I was just trying to avoid the constant array lookups on each invocation.